packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -61,6 +62,397 @@ enum _ScaffoldSlot {
+  /// {@end-tool}
+  ///
+  /// A less elegant but more expedient solution is to assign a [GlobalKey] to the
+  /// [ScaffoldMessenger], then use the `key.currentState` property to obtain the
+  /// [ScaffoldMessengerState] rather than using the [ScaffoldMessenger.of]
+  /// function. The [MaterialApp.scaffoldMessengerKey] refers to the root
+  /// ScaffoldMessenger that is provided by default.
+  ///
+  /// {@tool dartpad --template=freeform}
+  /// Sometimes [SnackBar]s are produced by code that doesn't have ready access
+  /// to a valid [BuildContext]. One such example of this is when you show a
+  /// SnackBar from a method outside of the `build` function. In these
+  /// cases, you can assign a [GlobalKey] to the [ScaffoldMessenger]. This
+  /// example shows a key being used to obtain the [ScaffoldMessengerState]
+  /// provided by the [MaterialApp].
+  ///
+  /// ```dart imports
+  /// import 'package:flutter/material.dart';
+  /// ```
+  /// ```dart
+  /// void main() => runApp(MyApp());
+  ///
+  /// class MyApp extends StatefulWidget {
+  ///   @override
+  ///  _MyAppState createState() => _MyAppState();
+  /// }
+  ///
+  /// class _MyAppState extends State<MyApp> {
+  ///   final GlobalKey<ScaffoldMessengerState> _scaffoldMessengerKey = GlobalKey<ScaffoldMessengerState>();
+  ///   int _counter = 0;
+  ///
+  ///   void _incrementCounter() {
+  ///     setState(() {
+  ///       _counter++;
+  ///     });
+  ///     if (_counter % 10 == 0) {
+  ///       _scaffoldMessengerKey.currentState.showSnackBar(const SnackBar(
+  ///         content: Text('A multiple of ten!'),
+  ///       ));
+  ///     }
+  ///   }
+  ///
+  ///   @override
+  ///   Widget build(BuildContext context) {
+  ///     return MaterialApp(
+  ///       scaffoldMessengerKey: _scaffoldMessengerKey,
+  ///       home: Scaffold(
+  ///         appBar: AppBar(title: Text('ScaffoldMessenger Demo')),
+  ///         body: Center(
+  ///           child: Column(
+  ///             mainAxisAlignment: MainAxisAlignment.center,
+  ///             children: <Widget>[
+  ///               Text(
+  ///                 'You have pushed the button this many times:',
+  ///               ),
+  ///               Text(
+  ///                 '$_counter',
+  ///                 style: Theme.of(context).textTheme.headline4,
+  ///               ),
+  ///             ],
+  ///           ),
+  ///         ),
+  ///         floatingActionButton: FloatingActionButton(
+  ///           onPressed: _incrementCounter,
+  ///           tooltip: 'Increment',
+  ///           child: Icon(Icons.add),
+  ///         ),
+  ///       ),
+  ///     );
+  ///   }
+  /// }
+  ///
+  /// ```
+  /// {@end-tool}
+  ///
+  /// If there is no [ScaffoldMessenger] in scope, then this will return null.
+  /// See also:
+  ///
+  ///  * [debugCheckHasScaffoldMessenger], which asserts that the given context
+  ///    has a [ScaffoldMessenger] ancestor.
+  static ScaffoldMessengerState of(BuildContext context) {
```

Comment:
Should this include a call to `debugCheckHasScaffoldMessenger` to assert with a nice message when it fails to find one it expected to find?

Also, a common pattern with these `of` calls is to include a `nullOk` optional `bool` that allows it to not assert and just return null.

I realize we do also have this pattern in some places of just returning null, but I guess I'm asking which is more appropriate for the use case: defaulting to throwing if it isn't found, or defaulting to not throwing and making the developer check for it? I think I'd lean toward the former, because it'll make them think about their choice.

Reply:
At one point I had it the way you describe, but it was massively breaking at the time. It's been re-worked a fair bit since then, so I think I should be able to add this back now. Thank you!

Reply:
I've applied your suggestions, PTAL. :)